### ISSUE TITLE
monitor directories on cifs mount does not see remote copied files

### ISSUE THREAD

**Issue by jtmoon79** [arvidn/libtorrent#3703]
**tl;dr** monitor directories under a _cifs_ mount type does not recognize remotely added `.torrent` files

I'm using qbittorrent-nox as the front-end.  I'm have a reasonable guess that [this Issue is due to libtorrent](https://github.com/arvidn/libtorrent/blob/a24afe8524d98fc5df27b8cda4a161a3c886ef8c/examples/client_test.cpp#L672) and not qBittorrent (i.e. the directory monitoring code _appears_ to originate in _libtorrent_, as best I could determine in 20 minutes of sleuthing).

### reproduction

Using `qbittorrent-nox`, I can add a list of paths to the _Monitored Folder_ entries (_Tools_ menu -> _Options_ -> _Downloads_ tab -> _Automatically add torrents from:_ sub-section).  However, a path that is mounted using _cifs_ does not seem to behave the same.

1. add a path to _Monitored Folder_ that is mounted using _cifs_<br/>
For example, path entry `/mnt/cifspath/monitor1`
2. check permissions allow reading and writing.  I tested the user running `qbittorrent-nox` could  `touch /mnt/cifspath/monitor1/test1`)
3. copy a `.torrent` file to that path (under any user permissions, didn't seem to matter) to the remotely shared and monitored directory.<br/>
I used Windows Explorer _from a different host_ and copy+pasted the `.torrent` file from a remote machine to a shared NAS path. e.g. from Windows host it was `\\nas\cifspath\monitor1` which appears to teh qbittorrent-nox host as path `/mnt/cifspath/monitor1/mytorrent.torrent`<br/>
**EXPECTED** `qbittorrent-nox` to add the `mytorrent.torrent` file to the ongoing session.<br/>
**ACTUAL** `qbittorrent-nox` does not react to the new file.<br/>
4. `touch` the `mytorrent.torrent` file (`touch /mnt/cifspath/monitor1/mytorrent.torrent`). <br/>
`qbittorrent-nox` does add the `mytorrent.torrent` file to the ongoing session and logs message:

        (N) 2019-03-10T18:49:16 - 'my torrent 12345' added to download list.

#### other small tests

- if a `.torrent` file is locally copied to the cifs mounted directory, then `qbittorrent-nox` does see it and process it, e.g. `cp /tmp/mytorrent.torrent /mnt/cifspath/monitor1/` results in log message `...added to download list.` and the removal of `mytorrent.torrent`.  This is expected.
- I also tested the `touch` as local user `root` and local user `qbtuser`. no difference in behavior (e.g. the qbittorrent-nox host processed the `/mnt/cifspath/monitor1/mytorrent.torrent` file and logged a message).
- I also tested re-saving the file `mytorrent.torrent` using notepad from the remote Windows Host.  This had no affect (`qbittorrent-nox` did not process the file at `/mnt/cifspath/monitor1/mytorrent.torrent`)
- I also tested other _Monitor_ paths under different mount types. Specifically, `vfat` and `ext4`.  These paths behaved as expected, e.g. `qbittorrent-nox` immediately processed and copied `.torrent` files that were copied to those local paths.
- I copied the `mytorrent.torrent` file using Windows Explorer to the remote NAS cifs share (`\\nas\cifspath\monitor1`), waited a long time.  (several hours).  `qbittorrent-nox` never notices the file.  (so it's not a delayed event issue or cache timeout issue or polling issue).

#### mount details

The `cifs` mounted path has mount options

    $ mount -v
    ...
    //nas/cifspath/monitor1 on /mnt/cifspath type cifs (rw,nosuid,nodev,noexec,relatime,vers=3.02,cache=strict,username=user1,domain=LOCAL,uid=1003,forceuid,gid=0,noforcegid,addr=192.168.1.10,file_mode=0755,dir_mode=0755,soft,nounix,serverino,nouser_xattr,mapposix,noperm,rsize=1048576,wsize=1048576,echo_interval=60,actimeo=60)

The CIFS service host (e.g. `//nas`) is running a Samba 4 release.

### overall

If the `.torrent` file is copied to the remote share from a different host then some filesystem event is not occurring on the Raspberry Pi host (where `qbittorrent-nox` and `libtorrent` are running).  This makes sense.  It probably infeasible for a CIFS host to broadcast all (or a subset of) filesystem events to all share-connected clients.  Or perhaps, the CIFS service (Samba) is capable of this but this option is not enabled.

It's likely not a bug in _libtorrent_.  I presumed this was a problem inherent in _inotify_ and remote filesystems.  However, github searches _repo:qbittorrent/qBittorrent inotify_ and _repo:arvidn/libtorrent inotify_  returned no results.

### workaround?

A simple workaround or "fix" would be to add an option for polling of _Monitor_ paths (e.g. periodically list path file entries or [something like this project](https://pypi.org/project/watchntouch/)).
Or, this Issue may entirely be in the domain of _qbittorrent/qBittorrent_ project code.

----

libtorrent version (or branch): _1.2.0 release_, qbittorrent-nox version _v4.1.5_ is the front-end.

platform/architecture: ARMv61 (Raspberry Pi B)

compiler and compiler version: gcc 6.3.0 20170516

**Comment by jtmoon79** [comment]
For anyone that has also run into this Issue, a workaround is to periodically `touch` files in the _Monitor_ directory.  This can be done with _cron_.

     # copy any another cron file to make sure cron file permissions are correct
     $ cp -av /etc/cron.d/john /etc/cron.d/touch-qbittorrent-drop
     # empty contents
     $ echo > /etc/cron.d/touch-qbittorrent-drop

set file contents of `/etc/cron.d/touch-qbittorrent-drop` to

    5 * * * *       root    find '/mnt/cifspath/monitor1' -mindepth 1 -maxdepth 1 -type f -name '*.torrent' -exec touch -c -- {} \; -exec logger -t "touch-qbittorrent-drop" -- "touched {}" \;

This will check the `qbittorrent-nox` _Monitor_ directory every 5 minutes.

**Comment by Chocobo1** [comment]
This is likely a problem in qbt, here you can see qbt tries to detect whether the path is pointing to a network mount, if it is then qbt will use the polling method:
https://github.com/qbittorrent/qBittorrent/blob/master/src/base/filesystemwatcher.cpp#L81

The function detecting the network mount:
https://github.com/qbittorrent/qBittorrent/blob/master/src/base/utils/fs.cpp#L345

The function scanning for new files:
https://github.com/qbittorrent/qBittorrent/blob/master/src/base/filesystemwatcher.cpp#L159

My guess the culprit is at `Utils::Fs::isNetworkFileSystem` probably it didn't match the magic number, or perhaps `dir.entryList()` in `FileSystemWatcher::processTorrentsInDir` didn't retrieve the files correctly (this would be an Qt issue).

**Comment by jtmoon79** [comment]
Thanks @Chocobo1 .  That looks like the most likely culprit.  I'll repost the Issue there.
